pytorch/training/docker/2.4/py3/cu124/Dockerfile.gpu

```diff
@@ -0,0 +1,573 @@
+    fsspec \
+    "urllib3<2" \
+    "idna>=3.7"\
+    "tqdm>=4.66.3" \
+    "requests>=2.32.0" \
+ && /opt/conda/bin/mamba clean -afy \
+ && rm -rf /etc/apt/sources.list.d/*
+
+# Install commond pip packages (in case of conda package is not available)
+RUN pip install --no-cache-dir opencv-python
+
+# For EFA, below flags are needed to install EFA on docker image
+#  -n, --no-verify       Skip EFA device verification and test
+#  -l, --skip-limit-conf Skip EFA limit configuration
+#  -k, --skip-kmod       Skip EFA kmod installation
+RUN mkdir /tmp/efa \
+ && cd /tmp/efa \
+ && curl -O https://s3-us-west-2.amazonaws.com/aws-efa-installer/aws-efa-installer-${EFA_VERSION}.tar.gz \
+ && tar -xf aws-efa-installer-${EFA_VERSION}.tar.gz \
+ && cd aws-efa-installer \
+ && apt-get update \
+ && ./efa_installer.sh -y --skip-kmod --skip-limit-conf --no-verify \
+ && rm -rf /tmp/efa \
+ && rm -rf /tmp/aws-efa-installer-${EFA_VERSION}.tar.gz \
+ && rm -rf /var/lib/apt/lists/* \
+ && apt-get clean
+
+ENV PATH="${OPEN_MPI_PATH}/bin:${EFA_PATH}/bin:${PATH}"
+ENV LD_LIBRARY_PATH=${OPEN_MPI_PATH}/lib/:${EFA_PATH}/lib/:${LD_LIBRARY_PATH}
+
+WORKDIR /root
+
+# Configure Open MPI and configure NCCL parameters
+RUN mv ${OPEN_MPI_PATH}/bin/mpirun ${OPEN_MPI_PATH}/bin/mpirun.real \
+ && echo '#!/bin/bash' > ${OPEN_MPI_PATH}/bin/mpirun \
+ && echo "${OPEN_MPI_PATH}/bin/mpirun.real --allow-run-as-root \"\$@\"" >> ${OPEN_MPI_PATH}/bin/mpirun \
+ && chmod a+x ${OPEN_MPI_PATH}/bin/mpirun \
+ && echo "hwloc_base_binding_policy = none" >> ${OPEN_MPI_PATH}/etc/openmpi-mca-params.conf \
+ && echo "rmaps_base_mapping_policy = slot" >> ${OPEN_MPI_PATH}/etc/openmpi-mca-params.conf \
+ && echo NCCL_DEBUG=INFO >> /etc/nccl.conf \
+ && echo NCCL_SOCKET_IFNAME=^docker0 >> /etc/nccl.conf
+
+# # Install HWLOC for ofi-aws-nccl
+# RUN mkdir /tmp/hwloc \
+# && cd /tmp/hwloc \
+# && wget https://download.open-mpi.org/release/hwloc/v${HWLOC_VERSION%.*}/hwloc-${HWLOC_VERSION}.tar.gz \
+# && tar -xf hwloc-${HWLOC_VERSION}.tar.gz \
+# && cd hwloc-${HWLOC_VERSION} \
+# && ./configure \
+# && make \
+# && make install \
+# && rm -rf /tmp/hwloc \
+# && rm -rf /var/lib/apt/lists/* \
+# && apt-get clean  
+
+
+# Install AWS-OFI-NCCL plugin from source
+RUN mkdir /tmp/aws-ofi-nccl \
+&& cd /tmp/aws-ofi-nccl \
+&& wget https://github.com/aws/aws-ofi-nccl/releases/download/v${AWS_OFI_NCCL_VERSION}-aws/aws-ofi-nccl-${AWS_OFI_NCCL_VERSION}-aws.tar.gz \
+&& tar -xvf aws-ofi-nccl-${AWS_OFI_NCCL_VERSION}-aws.tar.gz \
+&& cd aws-ofi-nccl-${AWS_OFI_NCCL_VERSION}-aws \
+&& ./autogen.sh \
+&& ./configure --prefix=/opt/aws-ofi-nccl \
+                --with-libfabric=${EFA_PATH} \
+                --with-cuda=${CUDA_HOME} \
+                --with-mpi=${OPEN_MPI_PATH} \
+                --disable-tests \
+&& make \
+&& make install \
+&& rm -rf /tmp/aws-ofi-nccl \
+&& rm -rf /var/lib/apt/lists/* \
+&& apt-get clean  
+
+
+# # fix libssl version dependency
+# # Install from source
+# RUN mkdir /tmp/libssl1/ \
```

Comment:
Why are we installing this from source?

Reply:
Ubuntu 22.04 doesn't use libssl1.1, it uses libssl3, but openssh has a dependency on it, that' why we try to do it. But it doesn't resolve the problem. We finally add a ENV LD_LIBRARY_PATH="/lib/x86_64-linux-gnu:${LD_LIBRARY_PATH}" to solve it.